Restore user preferences

TurtleShip opened the pull request:
## Fixes #445.


### Description of the changes:
**1) Do not set units to default values if they already have valid values** 
This fixes the actual issue. `UnitConverter::InitializeSelectedUnits()` ( this function resets all units to their default units if available for the current category ) gets called after `UnitConverterViewModel::RestoreUserPreferences()` ( this function restores user preferences ).
So Calculator has been restoring saved values, and then overriding the restored values with default values.
I modified `InitializeSelectedUnits()` so that we only initialize units only when they are not already set to valid units for the current category.

**2) Removed `m_isFirstTime`** 
I noticed that we are calling `RestoreUserPreferences()`  twice  when Calculator starts up, and the function is restoring the same value both times

The below happens when Calculator starts up
1) On startup, in `UnitConverterViewModel::InitializeView()`, `RestoreUserPreferences()` is called.
2) `RestoreUserPreferences()` in turn triggers `OnUnitChanged()`
3) During the first call to `OnUnitChanged()`, m_IsFirstTime is `True`, so we call `RestoreUserPreferences()` again while also setting `m_IsFirstTime` to `False`. 
4) `RestoreUserPreference()` again triggers `OnUnitChanged()`
5) During the second call to `OnUnitChanged()`,  m_IsFirstTime is `False`, so we call `SaveUserPreferences()`

I think we should only call `SaveUserPreferences()` inside `OnUnitChanged()` since we already restored user preferences during view initialization. I can't really think of a reason to restore units after view has been initialized. This led me to just delete `m_isFirstTime`.


### How changes were validated:
Manually tested that units and the current category are properly selected when you quit and start Calculator.

![GifMaker_20190414182150911](https://user-images.githubusercontent.com/3166423/56102706-88f73400-5ee3-11e9-8bbf-7a0c8e051a5c.gif)

![GifMaker_20190414183403644](https://user-images.githubusercontent.com/3166423/56102763-f0ad7f00-5ee3-11e9-99ef-3b932f587393.gif)

rudyhuyn commented on src/CalcManager/UnitConverter.cpp:
```diff
@@ -922,17 +922,25 @@ void UnitConverter::InitializeSelectedUnits()
     vector<Unit> curUnits = itr->second;
     if (!curUnits.empty())
     {
+        bool isFromRestored = find(curUnits.begin(), curUnits.end(), m_fromType) != curUnits.end();
```
NIT: `isFromUnitAvailable` would probably be easier to understand?

rudyhuyn commented on src/CalcManager/UnitConverter.cpp:
```diff
@@ -922,17 +922,25 @@ void UnitConverter::InitializeSelectedUnits()
     if (!curUnits.empty())
     {
+        bool isFromRestored = find(curUnits.begin(), curUnits.end(), m_fromType) != curUnits.end();
+        bool isToRestored = find(curUnits.begin(), curUnits.end(), m_toType) != curUnits.end();
```
NIT: rename `isFromUnitValid` and `isToUnitValid` + a comment explaining why we check them.

rudyhuyn replied:
small optimization: 
```
 bool isFromRestored = m_fromType == EMPTY_UNIT && find(curUnits.begin(), curUnits.end(), m_fromType) != curUnits.end();
```###

TurtleShip replied:
> NIT: rename `isFromUnitValid` and `isToUnitValid` + a comment explaining why we check them (units available for this region)

Thanks, @rudyhuyn !
I updated as suggested. Can you take a look at added comments please? I am afraid that you might have misunderstood why I was checking those values.

rudyhuyn replied:
The comment is right but another positive aspect of this check is that you will override units when they don't exist, many reasons possible:
- a unit was removed by a previous update (or moved to another category)
- the unit restored is not available for the current locale (user selected Pyeong, then changed region and started again the app)


-

TurtleShip replied:
Ah,, makes sense!
As for the second case, it shouldn't happen once #444 lands since the PR won't allow restoring values that aren't valid for the current locale in the first place.

rudyhuyn commented:
It really improves the user experience! I added 2 NIT comments, let me know what you think.
Thank you for taking care of that!

TurtleShip commented:
Hi, gentle bump on this.
It looks like I need a review from a reviewer with write access.
Could such a reviewer take a look please?

Thanks!

HowardWolosky commented:
Sorry for the delay, @TurtleShip.  I am looking at this, but I'm also looking through the archived history of the code around this.  The code wasn't exactly this way when it was initially written, and so I'm trying to review the history to see if there were other bugs/cases that were trying to be corrected along the way that we'd be regressing by taking this change.  I should have feedback on this PR next week.  Thanks for your patience.

TurtleShip commented:
Hi, @HowardWolosky 
I was wondering if you have updates on this 🙂

HowardWolosky commented:
>Hi, @HowardWolosky Howard Wolosky FTE
>I was wondering if you have updates on this 🙂

Hey there -- apologies for the delay.  Preparation for this year's BUILD conference took up a bit more time than anticipated.  I should have my investigation into the history around this code completed this week.  I appreciate your continued patience here.

TurtleShip commented:
Hi, a gentle bump on this... 🙂

HowardWolosky commented:
You're totally right on this @TurtleShip -- I apologize for the delay and radio silence here.  Other things came up preventing me from finishing that investigation.  I'll be prioritizing this soon and hope that we can get this merged in within the next two weeks.  I apologize again and appreciate your patience here.

TurtleShip commented on src/CalcViewModel/UnitConverterViewModel.cpp:
```diff
@@ -147,15 +147,13 @@ UnitConverterViewModel::UnitConverterViewModel(const shared_ptr<UCM::IUnitConver
 void UnitConverterViewModel::ResetView()
 {
     m_model->SendCommand(UCM::Command::Reset);
-    m_IsFirstTime = true;
```
Done :)

TurtleShip commented:
Hi, @HowardWolosky 

I've addressed your comments.
Thanks for taking time to look at my PR and I look forward to see it being merged to the master!
😄

rudyhuyn commented:
Good job Seulgi!
